# Five importannt ways for Imputing Missing Values

You cnn impute missing values using machine learning models.This process is known as data imputation and is commonly used in data preprocessing to handle the missing  or incomplete data .There are several methods and models you can use, depending  on the natire of your data and the missing values:

**1. `Simple Imputation Techniques:`**
   - **Mean/Median Imputation:** Replace missing values with the mean or median of the column.Suitable for numerical data.
   - **Mode Imputation:** Replace missing values with the mode(most frequent values)of the column.Useful for the categorical data.

**2. `k-Nearest Neighbors (KNN):`** This algrithm can be used to impute missing values based on  other variables in your dataset.
   
**3. `Regression Imputation:`** Use Regression model to predict the missing values based on other variables in your dataset.
   
**4. `Decision Tress and Random Forest:`** These can handle missing values inherently.Ther can also bee used to predict missing values based on the pattern  learned from the other data.
   
**5. ` Advanced Techniques:`**
   - **Multiple Imputation by Chained Equations:** This is more sophisticated technique that models each variables with missing values as a funtion of other variables in a round-robin fashion.
   - **Deep Learning Methods:** Neural Networks ,especially  autoencoders,can be effective in imputing  missing values in complex datasets.
  
**6. `Time Series Specific Methods:`** For time series data,you might use techniques like interpolation,forwardfillor backwoardill.
   

It's  important to choose the right method based on the type of the data, the pattern of missingness(e.g, at random,completely at random, or not at random) , and the amount of missing data.Additionally, it's crucial to understand that imputation can introduce bias or affect the distribution of your data, so it should be done with caution and an understanding of the potential implications.




# 1. Simple Impuation Techniques

## 1.1. Mean/Median Imputation 

Mean/median imputation replaces missing values with the mean or median of the column. This is a simple and effective method, but it has some limitations. For example, it reduces variance in the dataset, and it can lead to biased estimates if the missing values are not missing at random.

Let's see how to implement mean/median imputation in Python using the Titanic dataset.

### 1.1.1.Mean Imputation

In [1]:
import pandas as pd
import numpy  as np
import seaborn as sns
import matplotlib.pyplot as plt

#load the titanic dataset
df=sns.load_dataset('titanic')
df.head()



survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [2]:
#check the missing values
df.isnull().sum().sort_values(ascending=False)


deck           688
age            177
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

we can see that the `age` column has 177 missing values.Let's replace these missing values with the mean of the column:

In [3]:
#impute the missing values with mean
df['age']=df['age'].fillna(df['age'].mean())

#check the number of missing values in each column
df.isnull().sum().sort_values(ascending=False)

deck           688
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

we can see that the missing values of `age` column have been replaced with the mean of the column.

### 1.1.2. Median Imputation


Let's load the dataset and replace the missing values in the `age` column with the median of the column.

In [4]:
#load the datatset
data=sns.load_dataset('titanic')
data.isnull().sum().sort_values(ascending=False)


deck           688
age            177
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

In [5]:
#impute the missing values with median
data['age']=data['age'].fillna(df['age'].median())

#check the number of missing values in each column
data.isnull().sum().sort_values(ascending=False)

deck           688
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

### 1.2.3. Mode Imputation

Mode imputation replaces missing values with the mode (most frequent value)of the column. This is useful for imputing categorical columns, such as `Embarked` and `Embark_town` in the titanic dataset.

Let's see how to implement mode imputation in python using the titanic dataset

In [6]:
#load  the data
df=sns.load_dataset('titanic')

#check the missing values
df.isnull().sum().sort_values(ascending=False)


deck           688
age            177
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

In [8]:
#impute the missing  values with mode

df['embarked']=df['embarked'].fillna(df['embarked'].mode()[0])    # zero means that k phly mode  sy values ko replace krdo
df['embark_town']=df['embark_town'].fillna(df['embark_town'].mode()[0])

#check the missing values
df.isnull().sum().sort_values(ascending=False)

deck           688
age            177
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
embarked         0
class            0
who              0
adult_male       0
embark_town      0
alive            0
alone            0
dtype: int64

We can see tthat the missing values in the  `embark_town` and the `embarked` column have been replaced with the mode of column.

## 2. K-Nearest Neighbors(KNN)

KNN is a machine learning algorithm that  can be used for imputing the missing  values. it works by finding the most similiar data points to the one with the missing values based on the other available features. The missing values

In [9]:
#load the data
df=sns.load_dataset('titanic')

#chek the missing vlaues
df.isnull().sum().sort_values(ascending=False)

deck           688
age            177
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

In [11]:
#impute the missing values with KNN

from sklearn.impute import KNNImputer

#call the KNN imputer  with the  number of neighbors =4

imputer =KNNImputer(n_neighbors=4)

#fit the imputer
df['age']=imputer.fit_transform(df[['age']])

#check the number of missing values in each column
df.isnull().sum().sort_values(ascending=False)

deck           688
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64